src/spectrum/group.rs

```diff
@@ -692,47 +692,52 @@ mod mzsignal_impl {
         }
     }
 
-    /// Average a series of [`MultiLayerSpectrum`] together
-    pub fn average_spectra<C: CentroidLike + Default, D: DeconvolutedCentroidLike + Default>(
-        spectra: &[MultiLayerSpectrum<C, D>],
+    /// Average a series of [`SpectrumLike`] together. The supplied dx will be used to [`reprofile`]
+    /// centroid spectra as well as be the dx used for any profile spectra. The resulting
+    /// [`ArrayPair`] can be made into a [`BinaryArrayMap`] using `.into()`. Which in turn can be
+    /// used to create a new [`MultiLayerSpectrum`] or [`RawSpectrum`](crate::spectrum::RawSpectrum).
+    ///
+    /// Note: only available with feature `mzsignal`.
+    ///
+    /// # Warns
+    /// Warns you when the spectrum has `SignalContinuity::Profile` but not raw data array or when
+    /// the raw data arrays differ in dx with the supplied value.
+    pub fn average_spectra<S: SpectrumLike<C, D>, C: CentroidLike, D: DeconvolutedCentroidLike>(
+        spectra: impl IntoIterator<Item = S>,
         dx: f64,
-    ) -> ArrayPair<'_> {
-        let arrays: Vec<_> = spectra
-            .iter()
-            .map(|scan| {
-                if scan.signal_continuity() == SignalContinuity::Profile {
-                    if let Some(array_map) = scan.raw_arrays() {
-                        let mz = array_map.mzs().unwrap().to_vec();
-                        let inten = array_map.intensities().unwrap().to_vec();
-                        Some(ArrayPair::from((mz, inten)))
+    ) -> ArrayPair<'static> {
+        average_signal(
+            &spectra
+                .into_iter()
+                .flat_map(|spectrum| {
+                    if spectrum.signal_continuity() == SignalContinuity::Profile {
+                        if let Some(array_map) = spectrum.raw_arrays() {
+                            let mzs = array_map.mzs().unwrap().to_vec();
+                            let intensities = array_map.intensities().unwrap().to_vec();
+                            if let Some(spectrum_dx) = mzs.first().and_then(|start| {
+                                mzs.last().map(|end| (*end - *start) / (mzs.len() as f64))
+                            }) {
+                                if spectrum_dx != dx {
```

Comment:
You almost always want to reprofile or average to a higher resolution than the instrument measured to avoid having your grid not perfectly align with the instrument's uneven sampling of the m/z dimension, making this test unnecessary. 

Additionally, even profile data may have gaps in the m/z dimension, like those found in ion mobility datasets, so `(max - min) / n` may yield a value orders of magnitude too large. A closer approximation is is to compute the median or mode of the experimental m/z measures. Doing that efficiently is non-trivial, being `Θ(M_spectra * N_points_per_spectrum)` which is extra overhead for diagnostics that the caller may not need.

Reply:
Perfect I will remove this test then. I did not understand the use of the dx well enough, I was thinking every spectrum should have the exact same dx for this to work but now I understand that the given dx is just for reprofiling.